File: Problem solving with c++/Problem solving Level 2/Problem32.cpp
/*
Problem 32 |=================================================
Write a program to fill array with max size 100 with random numbers from
1t0 100, copy it to another array in reverse order and print it.

Input
10

Output:
Array 1 elements:
64 8 62 19 2 21 15 74 96 85

Array 2 elements after copying array 1 in reversed order: 
85 96 74 15 21 2 19 62 8 64
=============================================================
*/

#include <iostream>
#include <cstdlib>
#include <ctime>

using namespace std;

// Function to generate a random number between From and To (inclusive)
int RandomNumber(int From, int To)
{
    return rand() % (To - From + 1) + From;
}

// Function to fill an array with random numbers
void FillArrayWithRandomNumbers(int arr[100], int &arrLength)
{
    cout << "\nEnter number of elements:\n";
    cin >> arrLength; // User inputs the array length

    // Fill the array with random numbers between 1 and 100
    for (int i = 0; i < arrLength; i++)
        arr[i] = RandomNumber(1, 100);
}

// Function to print the elements of an array
void PrintArray(int arr[100], int arrLength)
{
    for (int i = 0; i < arrLength; i++)
        cout << arr[i] << " ";
    cout << "\n";
}

// Function to copy an array into another in reverse order
void CopyArrayInReverseOrder(int arrSource[100], int arrDestination[100], int arrLength)
{
    for (int i = 0; i < arrLength; i++)
        arrDestination[i] = arrSource[arrLength - 1 - i]; // Copying in reverse order
}

int main()
{
    // Seeds the random number generator using the current time
    srand((unsigned)time(NULL));

    // Declare an array and its length variable
    int arr[100], arrLength;

    // Fill the first array with random numbers
    FillArrayWithRandomNumbers(arr, arrLength);

    // Declare a second array for storing the reversed copy
    int arr2[100];

    // Copy elements of arr into arr2 in reverse order
    CopyArrayInReverseOrder(arr, arr2, arrLength);

    // Display elements of the first array
    cout << "\nArray 1 elements:\n";
    PrintArray(arr, arrLength);

    // Display elements of the second array after copying in reverse
    cout << "\nArray 2 elements after copy:\n";
    PrintArray(arr2, arrLength);

    return 0;
}

File: Problem solving with Python/Problem32.py
"""
Problem 32 |=================================================
Write a program to fill array with max size 100 with random numbers from
1t0 100, copy it to another array in reverse order and print it.

Input
10

Output:
Array 1 elements:
64 8 62 19 2 21 15 74 96 85

Array 2 elements after copying array 1 in reversed order: 
85 96 74 15 21 2 19 62 8 64
=============================================================
"""

import random

def random_number(from_val: int, to_val: int) -> int:
    """
    Generates a random number between 'from_val' and 'to_val' (inclusive).
    """
    return random.randint(from_val, to_val)

def fill_array_with_random_numbers(arr_length: int) -> list:
    """
    Fills an array with random numbers between 1 and 100.
    """
    return [random_number(1, 100) for _ in range(arr_length)]

def print_array(arr: list):
    """
    Prints the elements of an array.
    """
    print(" ".join(map(str, arr)))

def copy_array_in_reverse_order(arr: list) -> list:
    """
    Copies an array into another in reverse order.
    """
    return arr[::-1]

if __name__ == "__main__":
    # Seed the random number generator
    random.seed()

    # Read the number of elements
    arr_length = int(input("\nEnter number of elements:\n"))

    # Generate an array with random numbers
    arr = fill_array_with_random_numbers(arr_length)

    # Copy the array in reverse order
    arr2 = copy_array_in_reverse_order(arr)

    print("\nArray 1 elements:")
    print_array(arr)

    print("\nArray 2 elements after copy in reverse:")
    print_array(arr2)
